File: Leetcode_Medium/_36. Valid Sudoku/Solution.py
class Solution(object):
    def isValidSudoku(self, board):
        """
        :type board: List[List[str]]
        :rtype: bool
        """
        # Check rows
        for row in board:
            if not self.is_valid_unit(row):
                return False
        
        # Check columns
        for col in range(9):
            column = [board[row][col] for row in range(9)]
            if not self.is_valid_unit(column):
                return False
        
        # Check 3x3 squares
        for i in range(0, 9, 3):
            for j in range(0, 9, 3):
                square = [board[r][c] for r in range(i, i + 3) for c in range(j, j + 3)]
                if not self.is_valid_unit(square):
                    return False
        
        return True

    def is_valid_unit(self, unit):
        # Check if all elements in the unit are digits from 1 to 9
        digits = set()
        for cell in unit:
            if cell != '.':
                if cell in digits:
                    return False
                digits.add(cell)
        return True

File: Leetcode_Medium/_36. Valid Sudoku/Solution.java
import java.util.*;
class Solution {
    public boolean isValidSudoku(char[][] board) {
        // Check rows
        for (char[] row : board) {
            if (!isValidUnit(row))
                return false;
        }

        // Check columns
        for (int col = 0; col < 9; col++) {
            char[] column = new char[9];
            for (int row = 0; row < 9; row++) {
                column[row] = board[row][col];
            }
            if (!isValidUnit(column))
                return false;
        }

        // Check 3x3 squares
        for (int i = 0; i < 9; i += 3) {
            for (int j = 0; j < 9; j += 3) {
                char[] square = new char[9];
                int index = 0;
                for (int r = i; r < i + 3; r++) {
                    for (int c = j; c < j + 3; c++) {
                        square[index++] = board[r][c];
                    }
                }
                if (!isValidUnit(square))
                    return false;
            }
        }

        return true;
    }

    private boolean isValidUnit(char[] unit) {
        Set<Character> digits = new HashSet<>();
        for (char cell : unit) {
            if (cell != '.') {
                if (digits.contains(cell))
                    return false;
                digits.add(cell);
            }
        }
        return true;
    }
}
